src/derived/slot.rs

```diff
@@ -489,210 +492,137 @@ where
-                    } else {
-                        // We found this entry is valid. Update the
-                        // `verified_at` to reflect the current
-                        // revision.
-                        memo.revisions.verified_at = revision_now;
-                    }
-                }
+        };
 
-                QueryState::InProgress { .. } => {
-                    // Since we started verifying inputs, somebody
-                    // else has come along and started updated this
-                    // value. Just leave their marker alone and return
-                    // whatever `maybe_changed` value we computed.
-                }
+        let mut panic_guard =
+            PanicGuard::new(self.database_key_index, self, Some(old_memo), runtime);
+        let active_query = runtime.push_query(self.database_key_index);
 
-                QueryState::NotComputed => {
-                    // Since we started verifying inputs, somebody
-                    // else has come along and removed this value. The
-                    // GC can do this, for example. That's fine.
-                }
-            }
+        let memo = panic_guard.memo.as_mut().unwrap();
+        if memo
+            .revisions
+            .validate_memoized_value(db.ops_database(), revision_now, &active_query)
+        {
+            let maybe_changed = memo.revisions.changed_at > revision;
+            panic_guard.proceed();
+            maybe_changed
+        } else if memo.value.is_some() {
+            // We found that this memoized value may have changed
+            // but we have an old value. We can re-run the code and
+            // actually *check* if it has changed.
+            let StampedValue { changed_at, .. } =
+                self.execute(db, runtime, revision_now, active_query, panic_guard);
+            changed_at > revision
+        } else {
+            // We found that inputs to this memoized value may have chanced
+            // but we don't have an old value to compare against or re-use.
+            // No choice but to drop the memo and say that its value may have changed.
+            panic_guard.memo = None;
+            panic_guard.proceed();
+            true
         }
-
-        maybe_changed
     }
 
     /// Helper:
     ///
     /// When we encounter an `InProgress` indicator, we need to either
     /// report a cycle or else register ourselves to be notified when
-    /// that work completes. This helper does that; it returns a port
-    /// where you can wait for the final value that wound up being
-    /// computed (but first drop the lock on the map).
-    fn register_with_in_progress_thread(
+    /// that work completes. This helper does that. If a cycle is detected,
+    /// it returns immediately, but otherwise it blocks `self.database_key_index`
+    /// has completed and returns the corresponding result.
+    fn block_on_in_progress_thread<MutexGuard>(
```

Comment:
I wonder if `block_on_in_progress_query` would be a better name? It's a bit confusing that the `thread` might be this thread.